### PR TITLE
Add TableOfContentsPlugin

### PR DESCRIPTION
**Pull request by karam-qaoud**

### CODE REVIEW

**Review by zurfyx — COMMENTED** [review]
Almost there! Just some more nits

File: packages/lexical-playground/src/plugins/TableOfContentsPlugin.tsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+          }
+        }
+      }
+    }
+    lastScrollTop = st <= 0 ? 0 : st;
+  }
+  function onScroll(): void {
+    debounceFunction(scrollCallback, 10);
+  }
+  document.addEventListener('scroll', onScroll);
```
[the remark this answers is not in the document]
Comment (karam-qaoud) [reply packages/lexical-playground/src/plugins/TableOfContentsPlugin.tsx:118]: Nice catch! Will fix that.

[the remark this answers is not in the document]
Comment (karam-qaoud) [reply packages/lexical-playground/src/plugins/TableOfContentsPlugin.tsx:118]: It seems like I can't put `selectedIndex` inside of useEffect because scrollToNode needs to 'see' it.

File: packages/lexical-playground/src/plugins/TableOfContentsPlugin.tsx [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      timerId = setTimeout(func, delay);
+    }
+
+    function onScroll(): void {
+      debounceFunction(scrollCallback, 10);
+    }
+
+    document.addEventListener('scroll', onScroll);
+    return () => document.removeEventListener('scroll', onScroll);
+  });
```
Comment (zurfyx) [inline packages/lexical-playground/src/plugins/TableOfContentsPlugin.tsx:130]: Also, we may want to look into the 3 array optimization but we can do this in a separate PR
```suggestion
  }, [...]);
```
